### PR TITLE
Add inital PKCE support

### PR DESCRIPTION
**Pull request by orf**
Major ballache this one. 

I've got a working (or at least testable) branch here, forked from the 0.12.0 release that we currently run. IDK if its mergeable back into the parent project, but hey it should work for us.

### CODE REVIEW

File: oauth2_provider/forms.py [hunk 1]
```diff
@@ -8,3 +8,4 @@ class AllowForm(forms.Form):
     client_id = forms.CharField(widget=forms.HiddenInput())
     state = forms.CharField(required=False, widget=forms.HiddenInput())
     response_type = forms.CharField(widget=forms.HiddenInput())
+    pkce = forms.CharField(required=False, widget=forms.HiddenInput())
```
Comment (joealcorn) [inline oauth2_provider/forms.py]: Why is this a `CharField`? Wouldn't a boolean be more appropriate?

Comment (orf) [reply oauth2_provider/forms.py]: This field is where the hashed (or plainext) PKCE value is sent, if the field is present we assume that the whole auth flow is using PKCE and store the value in the grant.

Comment (joealcorn) [reply oauth2_provider/forms.py]: I'll re-read the spec and that article I originally read, but my understanding of PKCE is that this won't work. The value is random and will change with every authentication, so storing it here won't let more than 1 person auth at once.

Comment (joealcorn) [reply oauth2_provider/forms.py]: Actually, this is stored on Grant, I thought this was the `Application` model, nvm

File: oauth2_provider/tests/test_authorization_code.py [hunk 2]
```diff
@@ -908,6 +909,139 @@ def test_code_exchange_succeed_when_redirect_uri_match(self):
...
+
+        # retrieve a valid authorization code
+        authcode_data = {
+            'client_id': self.application.client_id,
+            'state': 'random_state_string',
+            'scope': 'read write',
+            'redirect_uri': 'http://example.it?foo=bar',
+            'response_type': 'code',
+            'allow': True,
+            'pkce': hashed_data
```
Comment (joealcorn) [inline oauth2_provider/tests/test_authorization_code.py]: I think we should call this `code_challenge`, that's what the spec says. The less deviation the better, how many unique oauth implementations have already been unleashed on the world?

File: oauth2_provider/tests/test_authorization_code.py [hunk 3]
```diff
@@ -908,6 +909,139 @@ def test_code_exchange_succeed_when_redirect_uri_match(self):
...
+        response = self.client.post(reverse('oauth2_provider:authorize'), data=authcode_data)
+        query_dict = parse_qs(urlparse(response['Location']).query)
+        authorization_code = query_dict['code'].pop()
+
+        # exchange authorization code for a valid access token
+        token_request_data = {
+            'grant_type': 'authorization_code',
+            'code': authorization_code,
+            'redirect_uri': 'http://example.it?foo=bar',
+            'code_challenge': 'secret'
```
Comment (joealcorn) [inline oauth2_provider/tests/test_authorization_code.py]: The key for the token request should be `code_verifier`: https://tools.ietf.org/html/rfc7636#section-4.5

**Review by joealcorn — APPROVED** [review]
I'm happy with this if you are @orf, good work

**Comment by orf** [comment]
lets do it. I'll make a PR on pie to include this

**Pull request merged**